doc/developer_guide/dependency.rst

```diff
@@ -0,0 +1,458 @@
+a single component.
+
+.. autoclass:: psyclone.core.access_info.Signature
+    :members:
+    :special-members: __hash__, __eq__, __lt__
+
+To collect access information in a `VariablesAccessInfo` object, the
+function `reference_accesses()` for the code region must be called:
+
+.. automethod:: psyclone.psyir.nodes.Node.reference_accesses
+
+.. autoclass:: psyclone.core.access_info.VariablesAccessInfo
+    :members:
+    :special-members: __str__
+
+This class collects information for each variable used in the tree
+starting with the given node. A `VariablesAccessInfo` instance can store
+information about variables in any arbitrary code, not only for a PSyIR
+node. You can pass it to more than one `reference_accesses()` function
+to add more variable access information, or use the `merge()` function to
+combine two `VariablesAccessInfo` objects into one. It is up to the user to
+keep track of which access information is stored in a `VariablesAccessInfo`
+instance.
+
+For each variable used an instance of
+`psyclone.core.access_info.SingleVariableAccessInfo` is created, which collects
+all accesses for that variable using `psyclone.core.access_info.AccessInfo`
+instances:
+
+.. autoclass:: psyclone.core.access_info.SingleVariableAccessInfo
+    :members:
+
+.. autoclass:: psyclone.core.access_info.AccessInfo
+    :members:
+
+Indices
+-------
+The `AccessInfo` class stores the original PSyIR node that contains the
+access, but it also stores the indices used in a simplified form, which
+makes it easier to analyse dependencies without having
+to analyse a PSyIR tree for details. Using the `component_indices` property of
+an `AccessInfo` object returns a list containing the indices used for
+each component. For example, an access like `a(i)%b(j,k)%c(l)` would
+return the list `[ [i], [j,k], [l] ]`. In case of non-array accesses,
+the corresponding index list will be empty, e.g. `a%b(j)%c` will
+have the component indices `[ [], [j], [] ]`, and a scalar `a` will just
+return `[ [] ]`.
+
+Each entry in this list of lists is still a PSyIR node for each index
+expression used. To find out details about an index expression, you can
+either analyse the tree, or use the variable access functionality again.
+Below is an example that shows how this is done to determine if an array expression
+contains a reference to a given variable `index_variable`. The variable
+`access_info` is an instance of `AccessInfo` and contains the information
+about one reference. The outer loop is over the indices used for each component,
+i.e. in case of `a(i)%b(j,k)%c(l)` it would loop over the three
+members of the list `[ [i], [j,k], [l] ]`. The inner loop then
+tests each index used, e.g. it would be `[i]` for the first index group,
+then `[j, k]` etc. Then the function `reference_accesses` is used to
+analyse the index expression.
+
+.. code-block:: python
+
+  for index_group in access_info.component_indices:
```

Comment:
Thinking about #1220, how much more code would be required here to make this example executable? e.g.

    psyir = fortran_reader.psyir_from_source("program my_prog\n....end program my_prog\n")
    ...